File: litedb-python/autoheal_smoke.py
"""
autoheal_smoke.py — proves the gossip failure detector AUTO-restores replication factor.

Scenario (4 nodes, RF 3, so reaping one genuinely re-replicates):
  1. Start 4 nodes; write 12 keys.
  2. Start the controller's failure detector and then **kill a node — with no manual controller call**.
  3. Gossip marks it dead; the controller's reconcile loop sees the majority verdict and, past the
     grace window, auto-fires remove_node(dead=True) → the dead node's shards re-replicate onto the
     survivors → RF back to 3 live replicas everywhere, data intact.

This closes the loop from "gossip detected a death" to "redundancy healed" with zero human action.
Run:  python autoheal_smoke.py
"""

import os

# 4 nodes so that reaping one actually moves a replica onto a survivor (RF 3 over 4 → 3 over 3).
os.environ.setdefault("LITEDB_CLUSTER_NODES", "4")

import shutil  # noqa: E402
import subprocess  # noqa: E402
import sys  # noqa: E402
import time  # noqa: E402
from collections import Counter  # noqa: E402

import _loader  # noqa: F401,E402
from cluster_client import ClusterClient  # noqa: E402
from cluster_config import DATA_ROOT, INITIAL_NODES, SHARDS  # noqa: E402
from controller import Controller  # noqa: E402


def spawn(nid: str):
    return subprocess.Popen([sys.executable, "node.py", nid],
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def wait_until(fn, timeout=40.0, what=""):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if fn():
            return True
        time.sleep(0.25)
    raise AssertionError("timeout: " + what)


def hosts_of(client):
    out = {}
    for st in client.status():
        if st.get("alive"):
            out[st["node"]] = {sh["group"] for sh in st.get("shards", [])}
    return out


def ready_leaders(client):
    m = {}
    for st in client.status():
        for sh in st.get("shards", []):
            if sh["role"] == "leader" and sh.get("ready"):
                m[sh["group"]] = sh["node"]
    return m


def main():
    n_shards = len(SHARDS)
    shutil.rmtree(DATA_ROOT, ignore_errors=True)
    procs = {n: spawn(n) for n in INITIAL_NODES}
    ctrl = Controller()
    try:
        client = ClusterClient()
        wait_until(lambda: len(ready_leaders(client)) == n_shards,
                   what=f"{n_shards} leaders on initial {len(INITIAL_NODES)} nodes")
        ctrl.broadcast_placement()
        print(f"started with {INITIAL_NODES} (RF 3)")

        for i in range(12):
            assert client.put(f"key{i}", f"val{i}").get("ok"), f"put key{i}"
        assert all(client.get(f"key{i}") == f"val{i}" for i in range(12))
        print("  wrote + read 12 keys")

        # The PD leader runs the failure detector autonomously — nothing to start. Kill a non-PD data
        # node (a clean data-node death; killing a PD member is covered by pd_failover_smoke).
        victim = INITIAL_NODES[-1]
        print(f"\nkilling {victim} — NO manual controller call; the PD failure detector must notice")
        procs[victim].kill()
        procs[victim].wait()

        wait_until(lambda: victim not in ctrl.active, what="controller starts reaping the dead node")
        print(f"  controller detected {victim} dead and is re-replicating (active now "
              f"{sorted(ctrl.active)})")

        # poll the OBSERVABLE replication state — remove_node frees `active` first and re-replicates
        # after, and Raft catch-up takes a moment, so wait for RF to actually be restored.
        def fully_healed() -> bool:
            h = hosts_of(client)
            if victim in h:
                return False
            c = Counter(s for ns in h.values() for s in ns)
            return len(c) == n_shards and all(c[s] == 3 for s in c)

        wait_until(fully_healed, what="RF re-replicated back to 3 on every shard")
        hosts = hosts_of(client)
        assert victim not in hosts, f"{victim} should host nothing: {hosts.get(victim)}"
        cnt = Counter(s for ns in hosts.values() for s in ns)
        assert all(cnt[s] == 3 for s in cnt), f"RF should be 3 on survivors: {dict(cnt)}"
        assert all(client.get(f"key{i}") == f"val{i}" for i in range(12)), "data intact after auto-heal"
        print(f"  RF restored to 3 across survivors {sorted(hosts)}; all 12 keys intact")

        print("\nAUTO-HEAL OK: a node died with zero manual action; gossip flagged it DEAD and the "
              "controller's failure detector re-replicated to restore RF, data preserved.")
    finally:
        ctrl.stop()
        for p in procs.values():
            if p.poll() is None:
                p.kill()
        for p in procs.values():
            p.wait()


if __name__ == "__main__":
    main()

File: litedb-java/src/main/java/com/litedb/cluster/AutoHealSmoke.java
package com.litedb.cluster;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * AutoHealSmoke — proves the gossip failure detector AUTO-restores replication factor (mirror of
 * autoheal_smoke.py). Start 4 nodes (RF 3), turn on the controller's failure detector, then KILL a
 * node with NO manual controller call. Gossip marks it dead; the reconcile loop sees the majority
 * verdict and re-replicates its shards onto survivors → RF back to 3, data intact.
 *
 * Run with 4 nodes:  LITEDB_CLUSTER_NODES=4 java com.litedb.cluster.AutoHealSmoke
 */
public class AutoHealSmoke {

    static Map<String, Process> procs = new LinkedHashMap<>();

    static Process spawn(String nid) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(System.getProperty("java.home") + "/bin/java",
                "-cp", System.getProperty("java.class.path"), "com.litedb.cluster.NodeServer", nid);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        Process p = pb.start();
        procs.put(nid, p);
        return p;
    }

    @SuppressWarnings("unchecked")
    static Map<String, java.util.Set<String>> hostsOf(ClusterClient client) {
        Map<String, java.util.Set<String>> out = new TreeMap<>();
        for (Map<String, Object> st : client.status()) {
            if (!Boolean.TRUE.equals(st.get("alive"))) continue;
            java.util.Set<String> shards = new java.util.HashSet<>();
            for (Object so : (List<Object>) st.get("shards")) {
                shards.add((String) ((Map<String, Object>) so).get("group"));
            }
            out.put((String) st.get("node"), shards);
        }
        return out;
    }

    @SuppressWarnings("unchecked")
    static int readyLeaders(ClusterClient client) {
        java.util.Set<String> withLeader = new java.util.HashSet<>();
        for (Map<String, Object> st : client.status()) {
            if (st.get("shards") == null) continue;
            for (Object so : (List<Object>) st.get("shards")) {
                Map<String, Object> sh = (Map<String, Object>) so;
                if ("leader".equals(sh.get("role")) && Boolean.TRUE.equals(sh.get("ready"))) {
                    withLeader.add((String) sh.get("group"));
                }
            }
        }
        return withLeader.size();
    }

    static Map<String, Integer> rfCount(Map<String, java.util.Set<String>> hosts) {
        Map<String, Integer> c = new HashMap<>();
        for (java.util.Set<String> shards : hosts.values()) {
            for (String s : shards) c.merge(s, 1, Integer::sum);
        }
        return c;
    }

    static void waitUntil(java.util.function.BooleanSupplier c, long ms, String what) throws Exception {
        long deadline = System.currentTimeMillis() + ms;
        while (System.currentTimeMillis() < deadline) { if (c.getAsBoolean()) return; Thread.sleep(250); }
        throw new AssertionError("timeout: " + what);
    }

    public static void main(String[] args) throws Exception {
        int nShards = ClusterConfig.SHARDS.size();
        Path root = Paths.get(ClusterConfig.dataRoot());
        if (Files.exists(root)) Files.walk(root).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
        Controller ctrl = new Controller(ClusterConfig.INITIAL_NODES, System.out::println);
        try {
            for (String n : ClusterConfig.INITIAL_NODES) spawn(n);
            ClusterClient client = new ClusterClient();
            waitUntil(() -> readyLeaders(client) == nShards, 25000,
                    nShards + " leaders on initial " + ClusterConfig.INITIAL_NODES.size() + " nodes");
            ctrl.broadcastPlacement();
            System.out.println("started with " + ClusterConfig.INITIAL_NODES + " (RF 3)");

            for (int i = 0; i < 12; i++) {
                if (!Boolean.TRUE.equals(client.put("key" + i, "val" + i).get("ok"))) throw new AssertionError("put key" + i);
            }
            for (int i = 0; i < 12; i++) if (!("val" + i).equals(client.get("key" + i))) throw new AssertionError("read key" + i);
            System.out.println("  wrote + read 12 keys");

            // The PD leader runs the failure detector autonomously — nothing to start. Kill a non-PD
            // data node (a clean data-node death; killing a PD member is covered by PdFailoverSmoke).
            String victim = ClusterConfig.INITIAL_NODES.get(ClusterConfig.INITIAL_NODES.size() - 1);
            System.out.println("\nkilling " + victim + " — NO manual controller call; the PD failure detector must notice");
            procs.get(victim).destroyForcibly();
            procs.get(victim).waitFor();

            waitUntil(() -> !ctrl.active().contains(victim), 40000, "controller starts reaping the dead node");
            System.out.println("  controller detected " + victim + " dead and is re-replicating (active now " + ctrl.active() + ")");

            // poll OBSERVABLE replication state — removeNode frees `active` first and re-replicates after.
            waitUntil(() -> {
                Map<String, java.util.Set<String>> h = hostsOf(client);
                if (h.containsKey(victim)) return false;
                Map<String, Integer> c = rfCount(h);
                if (c.size() != nShards) return false;
                for (int v : c.values()) if (v != 3) return false;
                return true;
            }, 40000, "RF re-replicated back to 3 on every shard");

            Map<String, java.util.Set<String>> hosts = hostsOf(client);
            if (hosts.containsKey(victim)) throw new AssertionError(victim + " should host nothing");
            for (int v : rfCount(hosts).values()) if (v != 3) throw new AssertionError("RF should be 3 on survivors: " + rfCount(hosts));
            for (int i = 0; i < 12; i++) if (!("val" + i).equals(client.get("key" + i))) throw new AssertionError("data lost after auto-heal");
            System.out.println("  RF restored to 3 across survivors " + hosts.keySet() + "; all 12 keys intact");

            System.out.println("\nAUTO-HEAL OK (Java): a node died with zero manual action; gossip flagged it DEAD "
                    + "and the controller's failure detector re-replicated to restore RF, data preserved.");
        } finally {
            ctrl.stop();
            for (Process p : procs.values()) p.destroyForcibly();
            for (Process p : procs.values()) p.waitFor();
        }
        System.exit(0);
    }
}
